crates/ditto-checker/src/typechecker/coverage/mod.rs

```diff
@@ -0,0 +1,515 @@
+                            let new_ideal = apply(&new_subst, ideal);
+                            let mut env = env.clone();
+                            env.extend(new_env);
+                            covered_by(
+                                &env,
+                                env_constructors,
+                                supply,
+                                &new_ideal,
+                                &clauses,
+                                not_covered,
+                            )
+                        })
+                }
+            }
+        } else {
+            let mut checked_clauses = vec![clause.clone()];
+            checked_clauses.extend(covered_by(
+                env,
+                env_constructors,
+                supply,
+                ideal,
+                remaining_clauses,
+                not_covered,
+            )?);
+            Ok(checked_clauses)
+        }
+    } else {
+        not_covered.push(ideal.clone());
+        Ok(clauses.to_vec())
+    }
+}
+
+type Subst = HashMap<FreshName, IdealPattern>;
+
+fn singleton_subst(fresh_name: FreshName, ideal_pattern: IdealPattern) -> Subst {
+    Subst::from([(fresh_name, ideal_pattern)])
+}
+
+fn has_subst(
+    supply: &mut Supply,
+    ideal: &IdealPattern,
+    clause_pattern: &ClausePattern,
+) -> Result<Option<Subst>> {
+    match (ideal, clause_pattern) {
+        (IdealPattern::Variable { var: fresh_name }, clause_pattern) => {
+            let subst = singleton_subst(
+                *fresh_name,
+                IdealPattern::from_clause(clause_pattern, supply),
+            );
+            Ok(Some(subst))
+        }
+        (
+            IdealPattern::Constructor {
+                constructor: ideal_constructor,
+                arguments: ideal_arguments,
+            },
+            ClausePattern::Constructor {
+                constructor: clause_constructor,
+                arguments: clause_arguments,
+                ..
+            },
+        ) => {
+            if ideal_constructor != clause_constructor {
+                return Ok(None);
+            }
+            if ideal_arguments.len() != clause_arguments.len() {
+                return Err(Error::MalformedPattern {
+                    ideal_arg_len: ideal_arguments.len(),
+                    clause_arg_len: clause_arguments.len(),
+                    malformed_pattern: clause_pattern.clone(),
+                });
+            }
+            let mut subst = Subst::new();
+            if ideal_arguments.is_empty() {
+                return Ok(Some(subst));
+            }
+            for (ideal_argument, clause_argument) in ideal_arguments.iter().zip(clause_arguments) {
+                if let Some(arg_subst) = has_subst(supply, ideal_argument, clause_argument)? {
+                    subst.extend(arg_subst)
+                } else {
+                    return Ok(None);
+                }
+            }
```

Comment:
Note (to self): got stuck on this bit for a while as the typeclass-y Haskell logic wasn't that obvious...

```haskell
mconcat <$> (sequence (zipWith hasSubst pats1 pats2))
```